HTML: better documentation for 'Void element' (#21195)
mdn/content: files/en-us/glossary/void_element/index.md

---
title: Void element
slug: Glossary/Void_element
tags:
  - CodingScripting
  - HTML
  - Intermediate
  - MathML
  - SVG
---

> A **void element** is an {{Glossary("element")}} from HTML, SVG, or MathML that **cannot** have any child nodes (i.e., nested elements or text nodes). Void elements only have a start tag; end tags must not be specified for void elements.

Author: done

In HTML, using a closing tag on a void element is invalid. For example, `<input type="text"></input>` is invalid HTML. Except for foreign elements (elements from SVG and MathML namespaces).

> The [HTML](https://html.spec.whatwg.org/multipage/), [SVG](https://www.w3.org/TR/SVG2/), and [MathML](https://www.w3.org/TR/MathML3/) specifications define very precisely what each element can contain. Many combinations have no semantic meaning, for example an {{HTMLElement("audio")}} element nested inside an {{HTMLElement("hr")}} element.

Reviewer: > Many combinations have no semantic meaning, for example an {{HTMLElement("audio")}} element nested inside an {{HTMLElement("hr")}} element.

Combinations of what? You can't nest anything in an hr, so this example confuses me. My preference is to delete the whole sentence, but maybe this is important. I just know know what you are trying to say.

Author: This para was in the original doc. I've edited the out the confusing example.

The void elements in HTML are as follows:

- {{HTMLElement("area")}}
- {{HTMLElement("base")}}
- {{HTMLElement("br")}}
- {{HTMLElement("col")}}
- {{HTMLElement("embed")}}
- {{HTMLElement("hr")}}
- {{HTMLElement("img")}}
- {{HTMLElement("input")}}
- {{HTMLElement("keygen")}}(HTML 5.2 Draft removed)
- {{HTMLElement("link")}}
- {{HTMLElement("meta")}}
- {{HTMLElement("param")}}
- {{HTMLElement("source")}}
- {{HTMLElement("track")}}
- {{HTMLElement("wbr")}}

## The curious case of self-closing tags

_Self-closing tags (`<tag />`) do not exist in HTML._\
However, foreign content is allowed to specify self-closing tags. In such case, if a start tag is marked as self-closing, then end tag must not be specified.

> If present, HTML parsers ignore the trailing `/` in start tags. Code formatters, like Prettier, take advantage of this fact and add a trailing `/` to the void tags to make them more readable.\

Reviewer 2: I suggest making the second sentence technically accurate while removing the positive value judgement claim ("readability"):
"Some code formatters use XML-inspired style and add a trailing `/` to tags for void elements."

There should probably be some text that says that `/` is part of the unquoted attribute value if not separated by space from the unquoted attribute value.

Author: done

> If present, HTML parsers ignore the trailing `/` in start tags. Code formatters, like Prettier, take advantage of this fact and add a trailing `/` to the void tags to make them more readable.\
> For example, `<input type="text">` to `<input type="text" />`

Reviewer: ```suggestion
If present, HTML parsers ignore the trailing `/` in start tags. Code formatters, like Prettier, take advantage of this fact and add a trailing `/` to the void tags to make them more readable. For example, code-formatters will convert `<input type="text">` to `<input type="text" />`.
```

---
title: Void element
slug: Glossary/Void_element
tags:
  - CodingScripting
  - HTML
  - Intermediate
---

A **void element** is an {{Glossary("element")}} in HTML that **cannot** have any child nodes (i.e., nested elements or text nodes). Void elements only have a start tag; end tags must not be specified for void elements.

> In HTML, using a closing tag on a void element is invalid. For example, `<input type="text"></input>` is invalid HTML. Except for foreign elements (elements from SVG and MathML namespaces).

Reviewer 2: The last sentence of this paragraph needs rework. It's unclear what "except" refers to. Elements in SVG and MathML for which child elements or text is non-conforming aren't "void elements". A more proper term would be "empty elements".

I suggest something like: "In contrast, empty elements in SVG and MathML may use an end tag instead of the XML self-closing syntax."

Author: done

> In HTML, using a closing tag on a void element is invalid. For example, `<input type="text"></input>` is invalid HTML. In contrast, empty elements in SVG and MathML may use an end tag instead of the XML self-closing syntax.

Reviewer 3: ```suggestion
In HTML, a void element must not have an end tag. For example, `<input type="text"></input>` is invalid HTML. In contrast, SVG or MathML elements that do have any child nodes may use an end tag instead of XML self-closing-tag syntax in their start tag.
```

Author: ```suggestion
In HTML, a void element must not have an end tag. For example, `<input type="text"></input>` is invalid HTML. In contrast, SVG or MathML elements that can not have any child nodes may use an end tag instead of XML self-closing-tag syntax in their start tag.
```

The [HTML](https://html.spec.whatwg.org/multipage/), [SVG](https://www.w3.org/TR/SVG2/), and [MathML](https://www.w3.org/TR/MathML3/) specifications define very precisely what each element can contain. So, some combinations of tags have no semantic meaning.

> Although a void element cannot have children in text serialized form (that is source code), child nodes can be added to the element in DOM using JavaScript. This is not a good practice as the outcome will not be reliable.

Reviewer 3: ```suggestion
Although there is no way to mark up a void element as having any children, child nodes can be added programmatically to the element in the DOM using JavaScript. But that is not a good practice, as the outcome will not be reliable.
```

> Although, a void element cannot have children in text serialized form (that is source code), child nodes can be added to the element in DOM using JavaScript. This is not a good practice as the outcome will not be reliable.

Reviewer: ```suggestion
Although a void element cannot have children in text serialized form (that is source code), child nodes can be added to the element in DOM using JavaScript. This is not a good practice as the outcome will not be reliable.
```

The void elements in HTML are as follows:

- {{HTMLElement("area")}}
- {{HTMLElement("base")}}
- {{HTMLElement("br")}}
- {{HTMLElement("col")}}
- {{HTMLElement("embed")}}
- {{HTMLElement("hr")}}
- {{HTMLElement("img")}}
- {{HTMLElement("input")}}
- {{HTMLElement("keygen")}}(HTML 5.2 Draft removed)
- {{HTMLElement("link")}}
- {{HTMLElement("meta")}}
- {{HTMLElement("param")}}
- {{HTMLElement("source")}}
- {{HTMLElement("track")}}
- {{HTMLElement("wbr")}}

## The curious case of self-closing tags

> _Self-closing tags (`<tag />`) do not exist in HTML._\

Reviewer 3: I guess the `\` character here won’t be rendered literally?

Author: It's the markdown way of adding `<br>`.

Reviewer 3: > It's the markdown way of adding `<br>`.

👍 TIL — I never knew, thanks

> However, foreign content is allowed to specify self-closing tags. In such case, if a start tag is marked as self-closing, then end tag must not be specified.

Reviewer 3: ```suggestion
However, the start tags of SVG and MathML elements that do have any child nodes are allowed to be marked as self-closing. In such cases, if an element's start tag is marked as self-closing, the element must not have an end tag.
```

Author: ```suggestion
However, the start tags of SVG and MathML elements that can not have any child nodes are allowed to be marked as self-closing. In such cases, if an element's start tag is marked as self-closing, the element must not have an end tag.
```

> If present, HTML parsers ignore the trailing `/` in start tags. Some code formatters, use this fact and add a trailing `/` to the void tags to make them similar to XML style tags. For example, code-formatters will convert `<input type="text">` to `<input type="text" />`.

Reviewer 2: The last sentence lacks the qualification "some".

Author: done

> If present, HTML parsers ignore the trailing `/` in start tags. Some code formatters, use this fact and add a trailing `/` to the void tags to make them similar to XML style tags. For example, some code-formatters will convert `<input type="text">` to `<input type="text" />`.

Reviewer 3: ```suggestion
If a trailing `/` (slash) character is present in the start tag of an HTML element, HTML parsers ignore that slash character. However, some code formatters add a trailing slash character to the start tags of void elements to make those start tags look similar to SVG and MathML start tags that are marked a self-closing. For example, some code formatters will convert `<input type="text">` to `<input type="text" />`.
```

Author: It about XHTML compatiblity and readability. Refer discussion thread: https://github.com/prettier/prettier/issues/5246#issuecomment-429029658
```suggestion
If a trailing `/` (slash) character is present in the start tag of an HTML element, HTML parsers ignore that slash character. However, some code formatters add a trailing slash character to the start tags of void elements to make them XHTML compatible and more readable. For example, some code formatters will convert `<input type="text">` to `<input type="text" />`.
```

> > **Note:** Without a space between unquoted attribute value and `/>`, the `/` becomes a part of the value. For example, the code `<img src=http://www.example.com/logo.svg/>` results in `src` attribute getting value `http://www.example.com/logo.svg/`, which makes the URL wrong.

Reviewer 3: ```suggestion
> **Note:** If a trailing `/` (slash) character in a start tag is directly preceded by unquoted attribute value — with no space between — the slash becomes a part of the attribute value rather than being discarded by the parser. For example, the markup `<img src=http://www.example.com/logo.svg/>` results in the `src` attribute having the value `http://www.example.com/logo.svg/` — which makes the URL wrong.
```
